server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -50,6 +58,47 @@ object SparkYarnApp extends Logging {
 
   private def getYarnPollInterval(livyConf: LivyConf): FiniteDuration =
     livyConf.getTimeAsMs(LivyConf.YARN_POLL_INTERVAL) milliseconds
+
+  private val appType = Set("SPARK").asJava
+
+  private val leakagedAppTags = new java.util.concurrent.ConcurrentHashMap[String, Long]()
+
+  private var sessionLeakageCheckTimeout: Long = _
+
+  private var sessionLeakageCheckInterval: Long = _
+
+  private val leakagedAppGCThread = new Thread() {
+    override def run(): Unit = {
+      while (true) {
+        if (!leakagedAppTags.isEmpty) {
+          // kill the app if found it and remove it if exceeding a threashold
+          val iter = leakagedAppTags.entrySet().iterator()
+          var isRemoved = false
+          val now = System.currentTimeMillis()
+          while(iter.hasNext) {
```

Comment:
I move `yarnClient.getApplications` to the outside of while loop so that we only call it one time.  LIVY-214 is only for hadoop-2.9, we can do it in another PR.